Tools/clinic/clinic.py

```diff
@@ -516,7 +516,13 @@ class PythonLanguage(Language):
     checksum_line = "#/*[{dsl_name} end generated code: {arguments}]*/"
 
 
-def permute_left_option_groups(l):
+ParamGroup = Iterable["Parameter"]
+ParamGenerator = Generator[tuple[ParamGroup, ...], None, None]
+
+
+def permute_left_option_groups(
+        l: Reversible[ParamGroup]
+) -> ParamGenerator:
```

Comment:
Several comments here:
1. I think you _might_ be overdoing it a bit with the type aliases from time to time -- here I think it maybe makes it harder to spot the bug in your annotation. You've annotated the function as returning `Generator[tuple[Iterable[Parameter], ...], None, None]`. A function that did that would look something like this:

   ```py
   def foo():
       while True:
           yield ([parameter], [parameter], [parameter])
   ```

   But I don't think that's what this function does. ISTM that you _give_ this function `Iterable[Iterable[X]]` and it returns a generator that yields `tuple[X, ...]`. So that would imply that the type hint should be `Generator[tuple[Parameter, ...], None, None]`
2. Having just told you to use `collections.abc.Generator` instead of `typing.Generator`... I actually wouldn't use `Generator` at all here, now that I take a closer look! The fact that this function is a generator function is an implementation detail that could easily go away in a later refactor. You're never `send`ing any values _into_ the generator, only getting values yielded _from_ the generator, so it's indistinguishable from any other iterator. As such, I'd go with a simpler and more readable type hint here for the return type -- either `Iterator[tuple[Parameter, ...]]` or `Iterable[tuple[Parameter, ...]]`. (I lean towards `Iterator` for generator-function return types, but there's not much difference between them.)
3. You've annotated the parameter with `Reversible[Iterable[Parameter]]`. That implies that the function would accept objects such as `[parameter, parameter, parameter]` -- when you call `reversed()` on a `list[parameter]`, you get an `Iterable[parameter]`, so this satisfies the annotation you've given. But that wouldn't be enough for this function -- this function requires an object which, when reversed, gives you an iterable of iterables of parameters. You _could_ annotate this with `Reversible[Iterable[Iterable[Parameter]]]`... or you could just do something more simple and readable (which should be fine for our purposes) -- `Sequence[Iterable[Parameter]]` :)

So, putting all of that together... here's how I'd annotate this function:

```py
def permute_left_option_groups(
    l: Sequence[Iterable[Parameter]]
) -> Iterator[tuple[Parameter, ...]]:
```

Reply:
(Note that I'm _not_ saying "don't use any type aliases" -- they can often greatly improve readability! But deeply nested type aliases can also _hurt_ readability by obscuring what the actual type is -- it can cut both ways; you need to take it on a case-by-case basis imo)

Reply:
In a previous version, I had four type aliases; that did not help 😆 Thanks for all the info! I'll have another go.